File: c_files/src/SpaceShip.c
/**
*
*@author Sude Andurman	sude.andurman@ogr.sakarya.edu.tr
*@since 03.05.2025
*<p>
*	Bilgisayar Mühendisliği 2. Sınıf 
*	Programlama Dilleri Prensipleri 1/C şubesi 
*</p>
*/

#include    "SpaceShip.h"
#include	"Person.h"
#include	"StringUtils.h" 


/* PRIVATE FUNCTIONS */
void		setState(SpaceShip this, const char* newState)
{
	/* CHECKING THE PARAMETERS */
	if (this == NULL || newState == NULL)
	{
		printf("Error: setState 'this' or 'newState' is NULL!\n");
		return;
	}

	/* REALLOC STATE */
	if (this->state != NULL) 
	{
		free(this->state);
	}
	this->state = strdup(newState);
}

void		updateState(SpaceShip this)
{
	/* CHECKING THE PARAMETERS AND VARIABLES */
	if (this == NULL || this->currentPlanet == NULL || this->destPlanet == NULL || this->launchDate == NULL)
	{
		printf("Error: updateState 'this' or its members are NULL!\n");
		return;
	}
	if (this->currentPlanet->currentDate == NULL)
	{
		printf("Error: updateState 'this' is NULL!\n");
		return;
	}

	/* UPDATE STATE */
	if (this->passengersCount == 0)
	{
		setState(this, "IMHA");
	}
	else if (this->launchDate->isEqual(this->launchDate, this->currentPlanet->currentDate))
	{
		setState(this, "Yolda");
	}
	else if (strcmp(this->state , "Yolda") == 0 && this->distanceLeft == 0)
	{
		setState(this, "Vardi");
	}
}

void		reduceDistanceLeft(SpaceShip this)
{
	/* CHECKING THE PARAMETERS */
	if (this == NULL || this->state == NULL)
	{
		printf("Error: reduceDistanceLeft 'this' or its state is NULL!\n");
		return;
	}

	/* REDUCE DISTANCE */
	if (strcmp(this->state, "Yolda") == 0 && this->distanceLeft > 0)
	{
		this->distanceLeft--;
	}
}
/* END OF PRIVATE FUNCTIONS */

SpaceShip	new_SpaceShip(const char* name, const char* currentPlanetName, const char* destPlanetName, int distanceLeft, int day, int month, int year)
{
	/* CHECKING THE PARAMETERS */
	if (name == NULL || currentPlanetName == NULL || destPlanetName == NULL)
	{
		printf("Memory allocation failed at new_SpaceShip\n");
		return (NULL);
	}

	/* MEMORY ALLOCATION FOR SPACESHIP*/
	SpaceShip this = (SpaceShip)malloc(sizeof(struct SPACESHIP));
	if (this == NULL)
	{
		printf("Memory allocation failed at new_SpaceShip\n");
		return (NULL);
	}

	/* MEMORY ALLOCATION FOR TIMELISTENER*/
	this->listener = new_ITimeListener();
	if (this->listener == NULL)
	{
		printf("Memory allocation failed for listener!\n");
		return NULL;
	}

	/* MEMORY ALLOCATION FOR MYLOCALTIME*/
	this->launchDate = new_MyLocalTime(day, month, year);
	if (this->launchDate == NULL)
	{
		printf("Memory allocation failed at new_SpaceShip MyLocalTime\n");
		free(this->listener);
		free(this);
		return (NULL);
	}

	/* SET INITIAL VALUES */
	this->name = strdup(name);
	this->currentPlanetName = strdup(currentPlanetName);
	this->destPlanetName = strdup(destPlanetName);
	this->state = strdup("Bekliyor");
	this->distanceLeft = distanceLeft;
	this->passengersCount = 0;

	/* LINKS */
	this->setInitials_Ship = &setInitials_Ship;
	this->calculateArrivalDate = &calculateArrivalDate;
	this->updatePassengers = &updatePassengers;
	this->delete_SpaceShip = &delete_SpaceShip;
	this->listener->onTimePass = &onTimePass_Ship; //interface implementation method
	/* END OF LINKS */

	return (this);
}

void		setInitials_Ship(const SpaceShip this, ArrayList planetList, ArrayList personList)
{
	/* CHECK THE PARAMETERS */
	if (this == NULL || this->currentPlanetName == NULL || this->destPlanetName == NULL)
	{
		printf("EROR: at setInitials_Ship \n");
		return;
	}

	/* SET INITIAL VALUES */
	this->currentPlanet = NULL;
	this->destPlanet = NULL;

	/* SET EXACT VALUES IF EXIST */
	if (planetList != NULL)
	{
		for (size_t i = 0; i < planetList->size; i++)
		{
			Planet planet = (Planet)(planetList->getByIndex(planetList, i));
			if (planet != NULL && planet->name != NULL)
			{
				if (equalsIgnoreCase(this->currentPlanetName, planet->name))
				{
					this->currentPlanet = planet;
				}
				if (equalsIgnoreCase(this->destPlanetName, planet->name))
				{
					this->destPlanet = planet;
				}

				/* IF THE VALUES THAT WE WANT FOUND BREAK THE LOOP*/
				if (this->currentPlanet != NULL && this->destPlanet != NULL)
				{
					break;
				}
			}
		}
	}else{
		printf("PlanetList is empty at setInitials_Ship\n");
	}

	/* SET THE PERSONLIST */
	if (personList != NULL)
	{
		this->allPersons = personList;
	}else{
    	printf("PersonList is empty at setInitials_Ship\n");
	}
}

void		calculateArrivalDate(const SpaceShip this)
{
	/* CHECKING THE PARAMETERS AND VARIABLES*/
	if (this == NULL || 
		this->currentPlanet == NULL || 
		this->destPlanet == NULL || 
		this->launchDate == NULL ||
		this->currentPlanet->currentDate == NULL ||
		this->destPlanet->currentDate == NULL)
	{
		printf("Error: calculateArrivalDate 'this' or its members are NULL!\n");
		return;
	}

	/* LOCAL VARIABLES */
	long daysBetween;
	long totalHours;
	long daysToAdd;
	this->arrivalDate = NULL;

	if ((this->currentPlanet != NULL) && (this->destPlanet != NULL))
	{
		// date copy to not effect the real date
		MyLocalTime destDateCopy;
		destDateCopy = this->destPlanet->currentDate->clone_MyLocalTime(this->destPlanet->currentDate);

		// get days between now and launch date
		daysBetween = this->currentPlanet->currentDate
							->daysBetween(this->launchDate,
											this->currentPlanet->currentDate);
		
		// calculate total hours to arrive to the destination
		totalHours = daysBetween * this->currentPlanet->hoursOfADay;
		totalHours += this->distanceLeft;

		// add days to destination planet's copy date to find arrival date
		daysToAdd = totalHours / this->destPlanet->hoursOfADay;
		destDateCopy->advanceDay(destDateCopy, daysToAdd);

		//set arrival date
		this->arrivalDate = destDateCopy;
	}

}

void		updatePassengers(const SpaceShip this)
{
	/* CHECKING THE PARAMETERS */
	if (this == NULL || this->allPersons == NULL)
	{
		printf("Error: updatePassengers 'this' is NULL!\n");
		return;
	}

	/* CALCULATE PASSANGERS COUNT*/
	int count = 0;
	for (int i = 0; i < this->allPersons->size; i++)
	{
		Person person = (Person)(this->allPersons->getByIndex(this->allPersons, i));
		if (person != NULL)
		{
			if (equalsIgnoreCase(person->currentVehicleName, this->name) && (person->isAlive(person)))
			{
				count++;
			}
		}else {
			printf("EROR: at updatePassangers person is null\n", i);
		}
	}
	
	/* SET PASSANGER COUNT */
	this->passengersCount = count;
}

void		delete_SpaceShip(SpaceShip this)
{
    if (this == NULL)
        return;
    
    if (this->name != NULL)
        free(this->name);

    if (this->currentPlanetName != NULL)
        free(this->currentPlanetName);
    
    if (this->destPlanetName != NULL)
        free(this->destPlanetName);
    
    if (this->state != NULL)
        free(this->state);
    
    if (this->listener != NULL)
        free(this->listener);

    if (this->launchDate != NULL)
        this->launchDate->delete_MyLocalTime(this->launchDate);
    
    if (this->arrivalDate != NULL)
        this->arrivalDate->delete_MyLocalTime(this->arrivalDate);
    
    free(this);
}

/* OVERRIDE METHOD */
void		onTimePass_Ship(const SpaceShip this)
{
	if (this == NULL)
	{
		printf("Error: onTimePass_Ship 'this' is NULL!\n");
		return;
	}

    updatePassengers(this);
    reduceDistanceLeft(this);
    updateState(this);
}

File: java_files/pdp_planet_travel_sim/src/app/SpaceShip.java
/**
*
*@author Sude Andurman	sude.andurman@ogr.sakarya.edu.tr
*@since 03.04.2025
*<p>
*	Bilgisayar Mühendisliği 2. Sınıf 
*	Programlama Dilleri Prensipleri 1/C şubesi 
*</p>
*/


package app;

/* IMPORTS */
import java.util.List;
/* END OF IMPORTS */

public class SpaceShip implements ITimeListener
{
	private String			name;				// TO SET NAME THAT WE GET FROM THE FILE
	private String			currentPlanetName;	// TO SET CURRENT PLANET THAT WE GET FROM THE FILE
	private String			destPlanetName;		// TO SET DESTINATION PLANET THAT WE GET FROM THE FILE
	private String			state;				// TO DETERMINE THE STATE
	private MyLocalTime		leavingDate;		// TO SET THE LEAVING DATE THAT WE GET FROM THE FILE
	private MyLocalTime		arrivalDate;		// TO SET THE ARRIVAL DATE THAT WE WILL CALCULATE
	private int 			distanceLeft;		// TO SET THE LEAVINF DATE THAT WE GET FROM THE FILE
	private int 			passengersCount;	// TO SET THE POPULATION OF THE SHIP
	private Planet			currentPlanet;		// TO MANIPULATE PLANET OBJECTS
	private Planet			destPlanet;			// TO MANIPULATE PLANET OBJECTS 
	private List<Person>	allPersons;			// TO CALCULATE HOW MANY PEOPLE CURRENTLY LOCATED IN THIS SHIP
	
	/* CONSTRUCTERS */
	
	public	SpaceShip(String name, String planetName, String destPlanetName, MyLocalTime leavingDate, int distance)
	{
		this.setName(name);
		this.setCurrentPlanet(planetName);
		this.setDestPlanet(destPlanetName);
		this.setLeavingDate(leavingDate);
		this.setDistanceLeft(distance);
		this.setArrivalDate(null);
		this.setState();
		this.setPassangersCount();
		Time.addListener(this);

	}
	
	/* END OF CONSTRUCTERS */
	
	
	/* SETTER METHODS */
	
	private void	setName(String name)
	{
		this.name = name;
	}
	private void	setCurrentPlanet(String planetName)
	{
		this.currentPlanetName = planetName;
	}
	
	private void	setDestPlanet(String planetName)
	{
		this.destPlanetName = planetName;
	}
	private void	setState()
	{
		this.state = "Bekliyor";
	}
	private void	setLeavingDate(MyLocalTime date)
	{
		this.leavingDate = date;
	}
	private void	setArrivalDate(MyLocalTime date)
	{
		if (date == null)
			this.arrivalDate = new MyLocalTime(1,1,1);
		else
			this.arrivalDate = date;
	}
	private void	setDistanceLeft(int hours)
	{
		this.distanceLeft = hours;
	}
	private void	setPassangersCount()
	{
		this.passengersCount = 0;
	}
	public void		setPlanetFromNames(List<Planet> planets)
	{
		// FIND THE PLANET OBJECT FROM THE PLANET NAME
		for (Planet planet : planets)
		{
			if (planet.getName().equalsIgnoreCase(currentPlanetName))
			{
				this.currentPlanet = planet;
			}
			if (planet.getName().equalsIgnoreCase(destPlanetName))
			{
				this.destPlanet = planet;
			}
		}
	}
	public void		setPersonList(List<Person> allPersons)
	{
		this.allPersons = allPersons;
	}
	
	/* END OF SETTER METHODS */
	
	/* GETTER METHODS */
	
	public String		getName()
	{
		return (name);
	}
	public String		getState()
	{
		return (state);
	}
	public String		getCurrentPlanetName()
	{
		return (currentPlanetName);
	}
	public String		getDestPlanetName()
	{
		return (destPlanetName);
	}
	public int			getDistanceLeft()
	{
		return (distanceLeft);
	}
	public int			getPassangersCount()
	{
		return (passengersCount);
	}
	public MyLocalTime	getLeavingDate()
	{
		return (leavingDate);
	}
	public MyLocalTime	getArrivalDate()
	{
		return (arrivalDate);
	}
	public Planet		getCurrentPlanet()
	{
		return (currentPlanet);
	}
	public Planet		getDestPlanet()
	{
		return (destPlanet);
	}
	
	/* END OF GETTER METHODS */

	/* METHODS */
	
	private void	updateState() {
		if (passengersCount == 0)
		{
			state = "İMHA";
			// IF THE SHIP IS NOT EXIST ANYMORE NO NEED TO BE RELATED TO THE TIME
			Time.removeListener(this);
		}
		else if (leavingDate.isEqual(currentPlanet.getCurrentDate()))
		{
			state = "Yolda";
		}
		else if (state.equals("Yolda") && distanceLeft == 0)
		{
			state = "Vardı";
		}
	}
	
	private void	reduceDistanceLeft()
	{
		if (state.equals("Yolda") && distanceLeft > 0)
		{
			distanceLeft--;
		}
	}
	
	public void		updatePassengers()
	{
		if (allPersons != null)
		{
			// IF THE PERSON IS ALIVE
			// AND
			// IF THE PERSON IS CURRENTLY IN THIS SHIP
			// COUNT TO CALCULATE POPULATION OF THE SHIP
			this.passengersCount = (int) allPersons.stream()
		            .filter(p -> p.isAlive() && p.getCurrentVehicle().equals(this.name))
		            .count();
		}
	}
	
	public void		calculateArrivalDate()
	{
		long daysBetween;	// DAYS BETWEEN CURRENT DATE AND LEAVING DATE
		long totalHours;	// CONVERT DAYS TO HOURS
		long daysToAdd;		// TO CALCULATE ARRIVAL DATE WHICH IS RELATED TO DESTINATION PLANET
		
		if (currentPlanet != null && destPlanet != null)
		{
			daysBetween = currentPlanet.getCurrentDate().daysBetween(leavingDate);
			totalHours = daysBetween * currentPlanet.getHoursOfADay();
			totalHours += distanceLeft;
		
			daysToAdd = totalHours / destPlanet.getHoursOfADay(); // CONVERT TOTAL HOURS TO DAYS
			setArrivalDate(new MyLocalTime(destPlanet.getCurrentDate().getCurrentDate().plusDays(daysToAdd)));
		}
	}
	
    public boolean	isValid(List<Planet> planets) // TO FIND THE PLANET THE SHIP CURRENTLY IS IN
    {
    	
    	boolean currentPlanetExists = planets.stream()
    	        .anyMatch(planet -> planet.getName().equalsIgnoreCase(this.currentPlanetName));
    	    
    	boolean destPlanetExists = planets.stream()
    	        .anyMatch(planet -> planet.getName().equalsIgnoreCase(this.destPlanetName));
    	
    	/* IF CURRENT PLANET AND DESTINATION PLANET ARE EXIST AT THE SAME TIME
    	 *  RETURN TRUE TO ADD THE SHIP TO THE UPDATED LIST */
    	return (currentPlanetExists && destPlanetExists);
    }
	
	@Override 
	public void		onTimePass()
	{
		// TRIGGER THE FUNCTIONS THAT RELATED TO THE TIME
		updatePassengers();
		reduceDistanceLeft();
		updateState();
	}
}


